### ISSUE TITLE
ValueAsHtmlString throws ArgumentNullException

### ISSUE THREAD

**Issue by ronaldbarendse** [umbraco/Umbraco.Forms.Issues#212]
The latest patch version 7.1.2 changed how the `ValueAsHtmlString` is returned, causing it to throw an `ArgumentNullException`.

Reproduction
------------

### Specifics
The following exception is thrown:
```
[ArgumentNullException: Value cannot be null.
Parameter name: source]
   System.Linq.Enumerable.FirstOrDefault(IEnumerable`1 source)
   Umbraco.Forms.Mvc.Models.FieldViewModel.get_ValueAsHtmlString()
   ASP._Views_Partials_Forms_Themes_default_Fieldtypes_FieldType_Textfield_cshtml.Execute()
```

### Steps to reproduce
Update to Umbraco Forms 7.1.2 and access `Model.ValueAsHtmlString` in a field type partial view.

### Expected result
Return the value or `null`/an empty `IHtmlString` if it's empty.

### Actual result
An `ArgumentNullException`.

**Comment by ronaldbarendse** [comment]
Just to add some context: the `ValueAsHtmlString` is used in the default views (e.g. `FieldType.Textfield.cshtml`), so probably **all sites are affected**.

Looking at the (decompiled) source, the implementation is changed from:

```c#
public IHtmlString ValueAsHtmlString
{
   get
   {
      return new HtmlString(HttpUtility.HtmlAttributeEncode(this.Value));
   }
}
```

```c#
public IHtmlString ValueAsHtmlString
{
   get
   {
      string str = HttpUtility.HtmlAttributeEncode((string)this.Values.FirstOrDefault<object>());
      return new HtmlString(str);
   }
}
```

I'm completely clueless why this change is introduced in the first place, then why it's not spotted in a code review and tests, plus finally why this change is automatically deployed on Umbraco Cloud (don't you validate multiple sites using a copy before pushing it to the rest)?

**Comment by ronaldbarendse** [comment]
Also looks like this related bug is still present: https://github.com/umbraco/Umbraco.Forms.Issues/issues/133

**Comment by nul800sebastiaan** [comment]
Hmm, while the code certainly needs some null checks, I can't actually reproduce the error on a test form, so we'll need to figure out why `Values` or `FirstOrDefault` is null for you.. odd!

**Comment by ronaldbarendse** [comment]
Thats really odd indeed! We do have a custom `Textfield`, but that just includes a single setting to add a custom input type:
```c#
using System;
using Umbraco.Forms.Core.Attributes;

/// <summary>
/// Custom text field type.
/// </summary>
/// <seealso cref="T:Umbraco.Forms.Core.Providers.FieldTypes.Textfield" />
public class Textfield : global::Umbraco.Forms.Core.Providers.FieldTypes.Textfield
{
	/// <summary>
	/// Gets or sets the type of the field.
	/// </summary>
	/// <value>
	/// The type of the field.
	/// </value>
	[Setting("Type", view = "dropdownlist", prevalues = "text,tel,url,email,date,month,week,time,datetime-local,number,color")]
	public string FieldType { get; set; }

	/// <summary>
	/// Initializes a new instance of the <see cref="Textfield" /> class.
	/// </summary>
	public Textfield()
		: base()
	{
		this.Id = new Guid("b569bedb-4966-4e68-986d-641eea40fa1d");
		this.Name = $"{this.Name} (field type)";
	}
}
```

**Comment by ronaldbarendse** [comment]
Just did some additional tests and `ValueAsHtmlString` always throws the error here!

Looking at the code, the `Values` property is `null` by default and is only set within `UmbracoFormsController.ResumeFromState()`, but only if the field exists in the form state... So it should be `null` when the form is rendered the first time and therefore throw the exception when accessing `ValueAsHtmlString`.

Besides the missing `null`-checks, the value is casted to a `string`, which is different than the other propertys (that converts it to a string using `ToString()`).

**Comment by nul800sebastiaan** [comment]
Yep, thanks I'm looking at the same code and what has changed and why. Values is null. Trying to find out where it comes from and why it would be null.

**Comment by ronaldbarendse** [comment]
It should also be noted that the values are [HTML **attribute** encoded](https://docs.microsoft.com/en-us/dotnet/api/system.web.httputility.htmlattributeencode), which encodes a subset of the normal [HTML encoding](https://docs.microsoft.com/en-us/dotnet/api/system.web.httputility.htmlencode). So `<tëst>` becomes `&lt;tëst>` instead of `&lt;t&#235;st&gt;`.

So for textareas, you should just use `Value` and let Razor HTML encode it (`<textarea>@Model.Value</textarea>`). Not sure why this property is obsoleted though...

For input values, attribute encoding can be used and because it does less, it's faster and the resulting HTML smaller. This implementation detail isn't obvious or properly documented **and** the default partial views don't always use the correct encoding (checkboxlist, dropdownlist and radiobuttonlist use `Value` for both attributes and label HTML and the textarea uses the attribute encoded value as HTML).

**Comment by ronaldbarendse** [comment]
For now, the workaround is to just use `ValueAsObject` in the partial views, so I don't get warnings about using the obsolete `Value` property and everything is HTML encoded...

**Comment by nul800sebastiaan** [comment]
I haven't had time to investigate why it's obsoleted while debugging this problem.

At this point I am suspecting something else custom in your solution. Could you try some things to help debug? 

- It could be that your custom fieldtype is a problem, if you switch to a regular textbox is it fixed?
- It could be a custom workflow, maybe temporarily remove that one and see if that fixes it
- It could also be a custom theme, can you switch to the default bootstrap theme while making sure that the files in the Views directory from `UmbracoForms.Files.7.1.2.zip` (https://our.umbraco.com/packages/developer-tools/umbraco-forms/) are all exactly the same in your solution

**Comment by ronaldbarendse** [comment]
Just pinpointed the issue to a custom MVC action filter that adds a honeypot field to all forms (that actually prevents most of the spam in contrast to the built-in field that is hidden using inline `display: none;` styling - thus easily detected by bots).

After initializing the property with `Values = new object[0]`, the error obviously disappeared. The fields that are configured within Umbraco Forms itself do get initialized to an empty array somewhere...

All my comments still apply however: the `null`-checks should be added again, instead of casting the value to a string, it should use `ToString()` and the correct encodings should be used in the partial views.

**Comment by nul800sebastiaan** [comment]
Sure, there were no actual null checks removed though, the previous code would also not check for null 

![image](https://user-images.githubusercontent.com/304656/63779630-28432380-c8e7-11e9-811f-87b110070b17.png)

I guess it wouldn't hurt to add `if(Values == null) return null;` though.

Encoding hasn't changed either.. which is.. interesting. I see no problem with `(string)`, it's a bit cheaper than `ToString()`, not that anyone will notice in this case.

**Comment by ronaldbarendse** [comment]
The `Value` property already contains the `null` check, so it **is** removed. And there's a **big** difference between casting to `string` and calling `ToString()` to returns its string representation:

```c#
var values = new List<object>() { 1 };
var valueToString = values.FirstOrDefault()?.ToString(); // "1"
var valueCast = (string)values.FirstOrDefault(); // InvalidCastException
```

As `HttpUtility.HtmlAttributeEncode()` (and `HtmlString`) accepts `null`  values, the property could just be rewritten to (using a [property expression](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/statements-expressions-operators/expression-bodied-members#read-only-properties) and [null-conditional operator](https://docs.microsoft.com/en-gb/dotnet/csharp/language-reference/operators/member-access-operators#null-conditional-operators--and-)):

```c#
public IHtmlString ValueAsHtmlString => new HtmlString(HttpUtility.HtmlAttributeEncode(this.Values?.FirstOrDefault()?.ToString()));
```

Or more verbose:
```c#
public IHtmlString ValueAsHtmlString
{
   get
   {
      var val = HttpUtility.HtmlAttributeEncode(this.Values?.FirstOrDefault()?.ToString());
      return new HtmlString(val);
   }
}
```

**Comment by nul800sebastiaan** [comment]
Thanks for the feedback! 👍 

This will be fixed in v7.1.3, a nightly is available with the fixes as listed above, if you want to help test. 

[UmbracoForms.Files.7.1.2-ci-20190828-001.zip](https://github.com/umbraco/Umbraco.Forms.Issues/files/3550130/UmbracoForms.Files.7.1.2-ci-20190828-001.zip)
